### PR TITLE
Added support for launching obfs4 proxies from srvreqq (incompatible with current configserver)

### PR DESCRIPTION
**Pull request by oxtoacart**
This requires a new request format.  Instead of requesting just a number, this now takes a JSON data structure with the following parameters:

```
{
    "id": 6,
    "obfs4_port": 62443,
    "srvq": "donyc3ox:srvq:manual",
}
```

id - this is the same as the number that we used to send.
obfs4_port - optional port for launching an obfs4 proxy instead of an https proxy.  This should always be 62443.
srvq - the queue to which the successful launch response is sent.  This used to always go to the same queue, but now client can specify it.

### CODE REVIEW

File: lib/vps_util.py [hunk 1]
```diff
@@ -226,7 +228,8 @@ def dc_by_cm(cm):
                             'vlpar1': 'ir',
                             'dosgp1': 'sea',
                             'dosfo1': 'sea',
-                            'vltok1': 'sea'}
+                            'vltok1': 'sea',
+                            'vllan1': 'etc',}
```
Comment (aranhoide) [inline lib/vps_util.py]: Whoops, missed this one, sorry!

**Comment by oxtoacart** [comment]
I tested this on Digital Ocean and Vultr by submitting a launch request from REDIS, like this:

```
RPUSH vllan1ox:srvreqq '{ "id": 3, "obfs4_port": 62443, "srvq": "vllan1ox:srvq:manual" }'
```

I tested with an HTTPS and an OBFS4 proxy.  The OBFS4 proxy ran into an issue with the http-proxy binary, but I'm resolving that separately.  Otherwise, everything seemed fine.

File: salt/cloudmaster/refill_srvq.py [hunk 2]
```diff
@@ -83,17 +84,20 @@ def kill_task(reqid):
                 print "Wat?"
                 break
         if len(pending) < MAXPROCS:
-            reqid, remover = reqq.next_job()
-            if reqid:
-                print "Got request", reqid
+            req_string, remover = reqq.next_job()
+            if req_string:
+                print "Got request", req_string
+                req = json.loads(req_string)
```
Comment (aranhoide) [inline salt/cloudmaster/refill_srvq.py]: In the old format, reqid would parse to a number.  Transition would be easier and safer if we supported this in an intermediary stage.  This should work:
```
if isinstance(req, int):
   req = {'id': req, 'srvq': QPREFIX + ':srvq'}
```

File: salt/cloudmaster/refill_srvq.py [hunk 3]
```diff
@@ -121,14 +125,17 @@ def get_lcs_name():
         pipe.set(CM + ':lcsserial', 1)
         pipe.execute()
         serial = 1
-    return 'fp-%s-%s-%03d' % (CM, date, serial)
+    type_prefix = 'obfs4' if 'obfs4_port' in req else 'https'
+    return 'fp-%s-%s-%s-%03d' % (type_prefix, CM, date, serial)
```
Comment (aranhoide) [inline salt/cloudmaster/refill_srvq.py]: This change breaks `vps_util.dc_by_name` (and some Pro server code).  I'm not sure what's the best solution here.  Thinking about it...

File: secret [hunk 4]
```diff
@@ -1 +1 @@
-Subproject commit e31ff8b245723d3ed69e5c0ec794f18e70884cc9
+Subproject commit 3aa7d884d3b5ac4107c59d7bc26d77f4a06f8999
```
Comment (aranhoide) [inline secret]: What secrets are added here?

**Comment by aranhoide** [comment]
@oxtoacart I'm done with the review.  The naming convention issue is thorny.  `fp-` is currently assumed in a lot of places, so removing the `-` there doesn't seem safe.

One thing we could do is to identify the cloudmaster/datacenter by index from the right.  Hopefully all current proxies will have been replaced before we need to break that too.  :P

File: lib/vps_util.py [hunk 5]
```diff
@@ -209,7 +209,9 @@ def cm_by_name(name):
     elif name.startswith('fp-jp-'):
         name = name.replace('jp', 'vltok1', 1)
 
-    return name.split('-')[1]
+    # We need to count from the right because we have HTTPS proxies both with
+    # and without the -https- part in their name.
+    return name.split('-')[-3]
```
Comment (aranhoide) [inline lib/vps_util.py]: Whoops, I meant to make this a sub-PR to merge into `obfs4launch` (or into master, doesn't matter a whole lot).

Anyway, please let me know whether this looks good to you.  If so, I'll submit an equivalent PR to the Pro server (which, being in Clojure, can't share this code).

File: secret (hunk 4 above)
Comment (oxtoacart) [inline secret]: Whatever's on master of too-many-secrets.

Comment (aranhoide) [inline secret]: OK, works for me.  I was just double-checking, should I have missed some new secret.

File: lib/vps_util.py (hunk 5 above)
Comment (oxtoacart) [inline lib/vps_util.py]: Yes, this looks good.

File: secret (hunk 4 above)
Comment (myleshorton) [inline secret]: Hey @oxtoacart does it always use 62443? Seems like that would make it easy
to block.

Comment (oxtoacart) [inline secret]: 62443 is internal. External is 443, which seems to be recommended for obfs4.

-- 
Cheers,
Ox

PGP Key - https://pgp.mit.edu/pks/lookup?op=get&search=ox@getlantern.org

-----------------------------------------------------------------------------------------

"I love people who harness themselves, an ox to a heavy cart,
who pull like water buffalo, with massive patience,
who strain in the mud and the muck to move things forward,
who do what has to be done, again and again."

- Marge Piercy

**Pull request merged**